examples/state_machine/states.py

```diff
@@ -0,0 +1,471 @@
+                else HandshakeFailedState
+        else:
+            next_state = HandshakingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = ReceivingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def reconnecting(self, event: events.HandshakeFailureEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.handshake_count + 1
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if self._context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = HandshakeReconnectingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def disconnect(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        return PNTransition(
+            state=HandshakeStoppedState,
+            context=self._context
+        )
+
+    def handshaking_success(self, event: events.HandshakeSuccessEvent, context: PNContext) -> PNTransition:
+        return PNTransition(
+            state=ReceivingState,
+            context=self._context
+        )
+
+
+class HandshakeReconnectingState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.DisconnectEvent.__name__: self.disconnect,
+            events.HandshakeReconnectGiveupEvent.__name__: self.give_up,
+            events.HandshakeReconnectSuccessEvent.__name__: self.success,
+            events.SubscriptionRestoredEvent.__name__: self.restored,
+            events.HandshakeReconnectFailureEvent.__name__: self.handshake_reconnect,
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+        }
+
+    def on_enter(self, context: Union[None, PNContext]):
+        self._context.update(context)
+        super().on_enter(self._context)
+        return effects.HandshakeReconnectEffect(self._context.channels, self._context.groups)
```

Comment:
I think it should also have number of attempts and reason of failure. The reason is for the cases when we have to stop retries and need to send the information about the failure reason

Reply:
What about the number of attempts?